core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcConnection.java

```diff
@@ -0,0 +1,552 @@
+          mConnectionOwner);
+      // Find handler for the request.
+      CopycatGrpcConnection.HandlerHolder handler = mHandlers.get(request.getClass());
+      if (handler != null) {
+        // Handle the request.
+        handler.getContext().executor().execute(() -> handleRequest(requestId, request, handler));
+      } else {
+        // Send fail response.
+        sendResponse(requestId, mContext,
+            new SerializationException("Unknown message type: " + request.getClass()));
+      }
+    } catch (SerializationException e) {
+      // Send fail response.
+      sendResponse(requestId, mContext, e);
+    }
+  }
+
+  /**
+   * Executes the request.
+   *
+   * @param requestId the request Id
+   * @param requestObject the request object
+   * @param handler registered handler for the request type
+   */
+  private void handleRequest(long requestId, Object requestObject,
+      CopycatGrpcConnection.HandlerHolder handler) {
+    // Call handler for processing the request.
+    CompletableFuture<Object> responseFuture = handler.getHandler().apply(requestObject);
+    // Send if there is a response.
+    if (responseFuture != null) {
+      responseFuture.whenComplete((response, error) -> {
+        // Action to send response based on future outcome.
+        Runnable responseAction = () -> {
+          if (error == null) {
+            sendResponse(requestId, mContext, response);
+          } else {
+            sendResponse(requestId, mContext, error);
+          }
+        };
+        // Make sure response is sent under catalyst context.
+        if (ThreadContext.currentContext() != null) {
+          // Current thread under catalyst context.
+          responseAction.run();
+        } else {
+          // Use originating context for dispatching the response.
+          mContext.executor().execute(responseAction);
+        }
+      });
+    }
+  }
+
+  /**
+   * Sends the response to target.
+   *
+   * Note: It sends an error response if given object it {@link Throwable}.
+   *
+   * @param requestId originating request Id
+   * @param context copycat thread context
+   * @param responseObject response object  to send
+   */
+  private void sendResponse(long requestId, ThreadContext context, Object responseObject) {
+    LOG.debug("Sending response of type: {} for request: {} from: '{}'",
+        responseObjectType(responseObject), requestId, mConnectionOwner);
+    // Create response message.
+    CopycatMessage.Builder messageBuilder =
+        CopycatMessage.newBuilder().setResponseHeader(CopycatResponseHeader.newBuilder()
+            .setRequestId(requestId).setFailed(responseObject instanceof Throwable));
+    // Serialize and embed response object if provided.
+    if (responseObject != null) {
+      messageBuilder.setMessage(UnsafeByteOperations
+          .unsafeWrap(context.serializer().writeObject(responseObject).array()));
```

Comment:
What are the requirements for using `unsafeWrap`, is it safe in our case?

Reply:
@bf8086 ?

Reply:
unsafe meaning that it will reuse the byte array instead of copy it to a private variable. In our case the byte array is not used anywhere else so it will not be a problem.